### ISSUE TITLE
ReCaptchaSettings.DetectionThreshold is not applied correctly

### ISSUE THREAD

**Issue by rjpowers10** [OrchardCMS/OrchardCore#17200]
### Describe the bug

My value of ReCaptchaSettings.DetectionThreshold is 5. I expect this to mean I can fail logging in 5 times and then on the 6th attempt I see the ReCaptcha.
* In OC 1.8.2 I actually have to fail 6 times before I see the ReCaptcha.
* In OC 2.1.2 I have to fail 7 times before I see the ReCaptcha.

In OC 2.1.2 it seems like the number of failed attempts went up by 1, but it seems like it was off in 1.8.2 as well. If the 1.8.2 behavior is correct then that's fine with me. But it does seem like it unintentionally changed in 2.1.2, which is my bigger concern.

### Orchard Core version

I noticed this in the process of updating from OC 1.8.2 to 2.1.2.

### To Reproduce

1. Create a new site with the blog recipe.
2. Enable the `OrchardCore.ReCaptcha.Users` feature.
3. Set the site key and secret key for ReCaptcha under Security > Settings > reCaptcha.
4. Log out.
5. Go to /login.
6. Repeatedly log in with dummy credentials.

### Expected behavior

The default setting of ReCaptchaSettings.DetectionThreshold is 5 and I'm using the default value.

In OC 1.8.2 I fail the log in process 5 times in a row. I would expect to see the ReCaptcha here, but I actually need to fail a 6th time for it to appear.

In this gif I click the "log in" button 5 times and fail, but it isn't until I click "log in" a 6th time that the ReCaptcha is displayed.

![1_8_2](https://github.com/user-attachments/assets/8f171f64-5a49-4be0-bf5c-2ea670aa1a4d)

In OC 2.1.2 it is similar, except I have to fail 7 times even though my threshold is still set to 5.
![2_1_2](https://github.com/user-attachments/assets/27a97d6b-7c44-4f3c-a383-64af7399daeb)

**Comment by Piedone** [comment]
I don't actually know why this worked differently before, since the logic is quite simple and implemented here, what didn't change: https://github.com/OrchardCMS/OrchardCore/blob/1c890daafa3f526f8ee6533f41f0833350da6b7a/src/OrchardCore/OrchardCore.ReCaptcha.Core/ActionFilters/Detection/IPAddressRobotDetector.cs#L38

So, the algorithm is that `DetectRobot()` (which checks `faultyRequestCount > _settings.DetectionThreshold`) is called when the login POST is handled, and `FlagAsRobot()`, which increments `faultyRequestCount`, after that, when the login fails. This is why you get 6 instead of 5.

You get 7 instead of 6 because `faultyRequestCount > _settings.DetectionThreshold` will still be `false` when the counter reaches `5`.

I only elaborated this in 10 times the characters the actual code is long, to show that for me it's really not clear how else this should be implemented, while agreeing that the configuration value of `DetectionThreshold` should work how you expect: after 5 failing logins, show captcha (i.e. at the 6th login).

Changing that `>` to `>=` is a trivial fix, but since this is the logic since 2018, I'm hesitant to just change it.

**Comment by rjpowers10** [comment]
I think I know why it increased from 6 to 7 in 2.1.2. It has to do with the refactoring to make the login form driver-based.

On the POST the form shape is built on line 108. This will validate the ReCaptcha and increment the counter if it fails.
https://github.com/OrchardCMS/OrchardCore/blob/4ab7c345f3c1f57e8fe4f7bcf173e86ad41aa5d8/src/OrchardCore.Modules/OrchardCore.Users/Controllers/AccountController.cs#L108

But if something about the POST fails (e.g. ReCaptcha validation fails) then the form shape is simply redisplayed as is with `View(formShape)`. The form needs to be rebuilt before it is redisplayed so we can check the counter again and potentially inject the ReCaptcha into the shape.

So to summarize I think there are two separate but related bugs.
1. The evaluation in `IPAddressRobotDetector` should be `>=` instead of `>`. I think this has been a bug since the beginning.
2. The form needs to be rebuilt before being redisplayed in `AccountController.LoginPOST`. This has been a bug since 2.0 when the login form was refactored to be driver-based.

**Comment by Piedone** [comment]
@MikeAlhayek opinion on #2?

**Comment by rjpowers10** [comment]
The model state is probably going to be lost if the form is rebuilt, so the entered username / email address will be lost. That's probably why it is the way it is now.

**Comment by MikeAlhayek** [comment]
### Self Notes

First observation, the `IPAddressRobotDetector` should not be using `IMemoryCache` as won't work in a distributed environment since there is no guarantee that the post request will be handed by the same machine where the memory is cached. We should use DistributedCache or the new HybridCache.

Second, it would be better for `IClientIPAddressAccessor` to return ValueTask instead of task with the IPAddress. The `IDetectRobots` should all be asyncronous calls and return `ValueTask`.


### Recommendation

On issue 1, I agree that the condition should be `>=` not `>` since we here start at 0 not 1.

On issue 2, I think the `ReCaptchaLoginFormDisplayDriver` should be replaced by a `ShapeTableProvider` so that the reCaptcha is not added using the display driver but using the shape table provider. I'll submit a PR for this fix.

**Comment by sebastienros** [comment]
> First observation, the IPAddressRobotDetector should not be using IMemoryCache

I don't quite agree, a local cache is still valid. It's just that every node has the cache. Not all caches need to be reused/shared. In this case it might be better to not issue a network call (when actually running distributed). There is no general truth in this case, it depends on the cost of the cached item.

**Comment by MikeAlhayek** [comment]
@sebastienros the problem here has nothing to so with distribution. This issue happens one a single node.

In regard to the `IPAddressRobotDetector`, we keep a count of how many attempts the user made to the server before showing the ReCaptcha challenge. The default settings indicate that the we show the challenge when the user submits the form 5 invalid attempts in a row, So if there are multiple nodes, and the user makes 4 attempts per server they will never see the challenge. So we should distribute the attempt count to ensure that the total requests to any node is 5.

**Comment by sebastienros** [comment]
@rjpowers10 we are now considering removing this threshold detection logic completely. The idea being that login already has a feature to limit the number of attempts, and rate limiting is an existing middleware that could also be used here (and anywhere else). The amount of changes and the fact we also need to maintain a shared cached to count (not thinking about eviction time) is too important for what we think is the actual usage: you are the first one to find it doesn't work even though this feature is not new.

**Comment by rjpowers10** [comment]
> The idea being that login already has a feature to limit the number of attempts, and rate limiting is an existing middleware that could also be used here (and anywhere else).

This is true but I'm unsure what this means for the ReCaptcha feature. I know .NET has a built in feature to lock an account after a number of failed attempts, but that feels like a separate topic from ReCaptcha.

Or are you thinking that ReCaptcha just becomes an on/off feature? No counter, the ReCatpcha is just shown every time (if enabled).

**Comment by MikeAlhayek** [comment]
> Or are you thinking that ReCaptcha just becomes an on/off feature? No counter, the ReCatpcha is just shown every time (if enabled).

exactly the idea.

**Comment by rjpowers10** [comment]
So you're only talking about removing the IP-based counter. Going forward, you can use ReCaptcha, or you can use the .NET lockout, or you can use both. ReCaptcha is either on or off with no detection features.

I'm starting to see why I'm confused. The IP-counter was only intended for logging in.
* In 1.8.2 the registration form shows the ReCaptcha on the first attempt, no counter check involved.
* In 2.1.2 registration is now checking the counter, but it shouldn't be. That's another bug.

Which makes sense now that I think about it more. Counting registration failures is kind of a weird concept to begin with.

Okay, removing the counter sounds good to me. I see where you're going with this.

**Comment by MikeAlhayek** [comment]
@Piedone are you also okay with removing the threshold and the counter from reCaptcha altogether?

**Comment by rjpowers10** [comment]
Are you thinking this would be a 2.1.x release? It's a breaking change...but there are also issues with it already.

**Comment by MikeAlhayek** [comment]
no. That would be a breaking change. 3.0

**Comment by rjpowers10** [comment]
Maybe this is worth a separate issue then but I think there is an issue with registration.
* 1.8.2 would show the ReCaptcha on the first attempt
* 2.1.2 is never showing the ReCaptcha because it is checking the counter but the counter also never gets incremented

**Comment by MikeAlhayek** [comment]
doesn't the reCaptcha always show now?

**Comment by rjpowers10** [comment]
Oh I see, the ReCaptcha tag helper has an "always show" mode that registration is using. Yes, you are correct. 2.1.2 registration is working. My mistake.

**Comment by Piedone** [comment]
> @Piedone are you also okay with removing the threshold and the counter from reCaptcha altogether?

Yep, good points.
